codalab/apps/chahub/tasks.py

```diff
@@ -0,0 +1,221 @@
+
+    try:
+        if include_deleted:
+            obj = Model.objects.all_objects().get(pk=pk)
+        else:
+            obj = Model.objects.get(pk=pk)
+    except Model.DoesNotExist:
+        raise ChahubException("Could not find {} with pk: {}".format(app_label, pk))
+    return obj
+
+
+@task(queue='site-worker')
+def send_to_chahub(app_label, model_name, pk, data, data_hash):
+    """
+    Does a post request to the specified API endpoint on chahub with the inputted data.
+    """
+    if not settings.CHAHUB_API_URL:
+        raise ChahubException("CHAHUB_API_URL env var required to send to Chahub")
+    if not settings.CHAHUB_API_KEY:
+        raise ChahubException("No ChaHub API Key provided")
+
+    obj = get_obj(app_label, model_name, pk)
+
+    try:
+        resp = _send(obj.get_chahub_endpoint(), data)
+    except requests.exceptions.RequestException:
+        resp = None
+
+    if resp and resp.status_code in (200, 201):
+        logger.info("ChaHub :: Received response {} {}".format(resp.status_code, resp.content))
+        obj.chahub_timestamp = timezone.now()
+        obj.chahub_data_hash = data_hash
+        obj.chahub_needs_retry = False
+    else:
+        status = getattr(resp, 'status_code', 'N/A')
+        body = getattr(resp, 'content', 'N/A')
+        logger.info("ChaHub :: Error sending to chahub, status={}, body={}".format(status, body))
+        obj.chahub_needs_retry = True
+    obj.save(send=False)
+
+
+@task(queue='site-worker')
+def delete_from_chahub(app_label, model_name, pk):
+    if not settings.CHAHUB_API_URL:
+        raise ChahubException("CHAHUB_API_URL env var required to send to Chahub")
+    if not settings.CHAHUB_API_KEY:
+        raise ChahubException("No ChaHub API Key provided")
+
+    obj = get_obj(app_label, model_name, pk, include_deleted=True)
+
+    url = "{}{}{}/".format(settings.CHAHUB_API_URL, obj.get_chahub_endpoint(), pk)
+    logger.info("ChaHub :: Sending to ChaHub ({}) delete message".format(url))
+
+    headers = {'X-CHAHUB-API-KEY': settings.CHAHUB_API_KEY}
+
+    try:
+        resp = requests.delete(url=url, headers=headers)
+    except requests.exceptions.RequestException:
+        resp = None
+
+    if resp and resp.status_code == 204:
+        logger.info("ChaHub :: Received response {} {}".format(resp.status_code, resp.content))
+        obj.delete()
+    else:
+        status = getattr(resp, 'status_code', 'N/A')
+        body = getattr(resp, 'content', 'N/A')
+        logger.info("ChaHub :: Error sending to chahub, status={}, body={}".format(status, body))
+        obj.chahub_needs_retry = True
+        obj.save(force_to_chahub=False)
+
+
+def batch_send_to_chahub(model, limit=None, retry_only=False):
+    qs = model.objects.all()
+    if retry_only:
+        qs = qs.filter(chahub_needs_retry=True)
+    if limit is not None:
+        qs = qs[:limit]
+
+    endpoint = model.get_chahub_endpoint()
+    data = [obj.get_chahub_data() for obj in qs if obj.get_chahub_is_valid()]
```

Comment:
because `get_chahub_data` returns `[{data:data}]` this will cause this array to look like: `[[{data: data}], [{data: data}]]` which will not be readable by chahub. (at least, this looks to be the case on competitions and datasets)

Reply:
that's true, how is v2 handling this? copied that code from there

Reply:
remove array from each `get_chahub_data` and wrap it in a list in the save method

Reply:
✔️